docs/src/index.md

```diff
@@ -44,52 +17,321 @@ Autodiff (AD) tools roughly work by reducting a program down to simple parts tha
+ - Wirtinger: it is complex. The docs need to be better. [Read the links in this issue](https://github.com/JuliaDiff/ChainRulesCore.jl/issues/40).
+ - Casted: it implements broadcasting mechanics. See [#10](https://github.com/JuliaDiff/ChainRulesCore.jl/issues/10)
+ - InplacableThunk: it is like a Thunk but it can do `store!` and `accumulate!` inplace.
+
+
+ -------------------------------
+## Example of using ChainRules directly.
+
+While ChainRules is largely intended as a backend for Autodiff systems it can be used directly.
+(Infact this can be very useful if you can constraint the code you need to differnetiate to only use thing that have rules defined for.
+This was once how all neural network code worked.)
+
+Using ChainRules directly also helped get a feel for it.
+
+
+```julia
+using ChainRules
+
+function foo(x)
+    a = sin(x)
+    b = 2a
+    c = asin(b)
+    return c
+end;
+
+###
+# Find dfoo/dx via rrules
+
+# First the forward pass, accumulating rules
+x=3;
+a, a_pb = rrule(sin, x);
+b, b_pb = rrule(*, 2, a);
+c, c_pb = rrule(asin, b)
+
+# Then the backward pass calculating gradients
+c̄ = 1;
+_, b̄ = c_pb(extern(c̄));
+_, _, ā = b_pb(extern(b̄));
+_, x̄ = a_pb(extern(ā));
+extern(x̄)
+# -2.0638950738662625
+
+###
+# Find dfoo/dx via frules
+
+# Unlike rrule can interleave evaluation and derivative evaluation
+x=3;
+ẋ=1;
+nofields = NamedTuple();
+
+a, a_pf = frule(sin, x);
+ȧ = a_pf(nofields, extern(ẋ));
+
+b, b_pf = frule(*, 2, a);
+ḃ = b_pf(nofields, 0, extern(ȧ));
+
+c, c_pf = frule(asin, b);
+ċ = c_pf(nofields, extern(ḃ));
+extern(ċ)
+# -2.0638950738662625
+
+###
+# Find dfoo/dx via finite-difference
+using FiniteDifferences
+central_fdm(5,1)(foo, x)
+# -2.0638950738670734
+
+###
+# Via ForwardDiff.jl
+using ForwardDiff
+ForwardDiff.derivative(foo, x)
+# -2.0638950738662625
+
+###
+# Via Zygote
+using Zygote
+Zygote.gradient(foo, x)
+# (-2.0638950738662625,)
 ```
+
+
+ -------------------------------
+
+
+
+## On writing good rrule / frules
+
+### Return Zero or One
+rather tan `0` or `1`
+or even rather than `zeros(n)`, `ones(m,n)`
+
+### Use thunks appropriately:
+
+If work is only required for 1 of the returned differentials it should be wrapped in a `@thunk` (potentially using a begin-end block)
+
+If there are multiple return values, almost always their should be computation wrapped in a `@thunk`s
+
+Don’t wrap variables in thunks, wrap the computations that fill those variables in thunks: Eg:
+Write:
+```julia
 ∂A = @thunk(foo(x))
 return ∂A
 ```
-                Not:
-```
+Not:
+```julia
 ∂A = foo(x)
 return @thunk(∂A)
 ```
 In the bad example `foo(x)` gets computed eagerly, and all that the thunk is doing is wrapping the already calculated result in a function that returns it.
 
+### Becareful of using Adjoing when you mean Transpose
 
+Rember for complex numbers `a'` (i.e. `adjoint(a)`) takes the complex conjugate. Instead you probably want `transpose(a)`.
 
+While there are arguments that for reverse-mode
```

Comment:
I'm not a big fan of the wording here, since it implies that one approach would be more correct than the other, when it's actually just a matter of which definition you use. I've just made a PR to explain this a bit better in the Zygote docs in FluxML/Zygote.jl#328, feel free to reference this here.